libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1065,6 +1129,198 @@ string YulUtilFunctions::clearStorageArrayFunction(ArrayType const& _type)
 	});
 }
 
+string YulUtilFunctions::copyArrayFromMemoryToStorage(ArrayType const& _fromType, ArrayType const& _toType)
+{
+	solAssert(_fromType.location() == DataLocation::Memory, "");
+
+	string functionName = "copy_array_memory_to_storage_from_" + _fromType.identifier() + "_to_" + _toType.identifier();
+	return m_functionCollector.createFunction(functionName, [&](){
+		Whiskers templ(R"(
+			function <functionName>(slot, value) {
+				let length := <arrayLength>(value)
+				<?isToDynamic>
+					<resizeArray>(slot, length)
+				</isToDynamic>
+
+				let memoryOffset := <dataLocation>(value)
+
+				let elementSlot := <toDataLocation>(slot)
+				<?isBytes>
+					if lt(length, 32) {
+						elementSlot := slot
+					}
+				</isBytes>
+				let elementOffset := 0
+
+				for { let i := 0 } lt(i, length) {i := add(i, 1)} {
+					let <elementValues> := <readFromMemory>(memoryOffset)
+					<?isBytes>
+						<updateStorageValue>(elementSlot, sub(31, elementOffset), <elementValues>)
+					<!isBytes>
+						<updateStorageValue>(elementSlot<?hasOffset>, elementOffset</hasOffset>, <elementValues>)
+					</isBytes>
+					memoryOffset := add(memoryOffset, <stride>)
+
+					<updateSlotAndOffset>
+				}
+			}
+		)");
+		templ("functionName", functionName);
+		templ("value", suffixedVariableNameList("value_", 0, _fromType.sizeOnStack()));
+		templ("isToDynamic", _toType.isDynamicallySized());
+		templ("isBytes", _toType.isByteArray());
+		if (_toType.isDynamicallySized())
+			templ("resizeArray", resizeDynamicArrayFunction(_toType));
+		templ("dataLocation", arrayDataAreaFunction(_fromType));
+		templ("toDataLocation", arrayDataAreaFunction(_toType));
+		templ("arrayLength",arrayLengthFunction(_fromType));
+		templ("stride", to_string(_fromType.memoryStride()));
+		templ("readFromMemory", readFromMemory(*_fromType.baseType()));
+		templ("hasOffset", _toType.baseType()->isValueType());
+		templ("elementValues", suffixedVariableNameList(
+				"elementValue_",
+				0,
+				_fromType.baseType()->stackItems().size()
+		));
+		templ("updateStorageValue", updateStorageValueFunction(*_fromType.baseType(), *_toType.baseType()));
+		templ(
+			"updateSlotAndOffset",
+			Whiskers(R"(
+				<?multipleItemsPerSlot>
+					elementOffset := add(elementOffset, <storageStride>)
+					if gt(elementOffset, sub(32, <storageStride>)) {
+						elementOffset := 0
+						elementSlot := add(elementSlot, 1)
+					}
+				<!multipleItemsPerSlot>
+					elementSlot := add(elementSlot, <storageSize>)
+					elementOffset := 0
+				</multipleItemsPerSlot>
+			)")
+			("multipleItemsPerSlot", _toType.storageStride() <= 16)
+			("storageStride", to_string(_toType.storageStride()))
+			("storageSize", _toType.baseType()->storageSize().str())
+			.render()
+		);
+		return templ.render();
+	});
+}
+
+string YulUtilFunctions::copyArrayFromCalldataToStorage(ArrayType const& _fromType, ArrayType const& _toType)
```

Comment:
It may be easier to combine copying from calldata and copying from memory than it is to combine regular arrays and byte arrays :)

Reply:
That was my idea as well. Also byte arrays will need #9966 first :)
Starting rebase now

Reply:
I went with removing byte array implementation from this PR as it needs #9966 first. It will probably end up in its own function. I left calldata and memory cases separate for now. If you think it should be combined I can merge them.